### ISSUE TITLE
in jetty12 beta partial http data request can get stuck now

### ISSUE THREAD

**Issue by andresluuk** [eclipse/jetty.project#10123]
**Jetty version(s)**
Built jetty 12 from source and made a profile with:
java -jar start.jar --add-module=server,http,deploy,ee10-deploy,ee10-jsp,resources,ee10-jstl,client

**Java version/vendor** `(use: java -version)`
OpenJDK Runtime Environment Temurin-17.0.4+8 (build 17.0.4+8)

**OS type/version**
Windows

**Description**
I uploaded a war example (the zip should be renamed to war).
Anyway. I am making a HttpURLConnection requesti into a immage with a range:
connection.setRequestProperty("Range", "bytes=1100-2300,36000-41000,45012-45099" );
And now reading the stream never finishes.
From the servers side I can see that there is a thread stuck at:
```
Thread: "qtp1581267786-86" Id=86 RUNNABLE daemon=false
	at java.base@17.0.4/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at java.base@17.0.4/sun.nio.ch.FileDispatcherImpl.read(FileDispatcherImpl.java:54)
	at java.base@17.0.4/sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:330)
	at java.base@17.0.4/sun.nio.ch.IOUtil.read(IOUtil.java:284)
	at java.base@17.0.4/sun.nio.ch.IOUtil.read(IOUtil.java:273)
	at java.base@17.0.4/sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:232) - locked java.lang.Object@206e4ed1
	at org.eclipse.jetty.io.content.PathContentSource.read(PathContentSource.java:164)
	at org.eclipse.jetty.http.MultiPartByteRanges$PathContentSource.read(MultiPartByteRanges.java:160)
	at org.eclipse.jetty.io.content.PathContentSource.read(PathContentSource.java:138)
	at org.eclipse.jetty.http.MultiPart$AbstractContentSource.read(MultiPart.java:761)
	at org.eclipse.jetty.io.internal.ContentCopier.process(ContentCopier.java:52)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:243)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224)
	at org.eclipse.jetty.io.Content.copy(Content.java:91)
	at org.eclipse.jetty.io.Content.copy(Content.java:66)
	at org.eclipse.jetty.server.ResourceService.sendData(ResourceService.java:694)
	at org.eclipse.jetty.server.ResourceService.doGet(ResourceService.java:198)
	at org.eclipse.jetty.ee10.servlet.DefaultServlet.doGet(DefaultServlet.java:498)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:527)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614)
```
It was passing with jetty10 and 11.
[pkg-http-partialContent.zip](https://github.com/eclipse/jetty.project/files/12078802/pkg-http-partialContent.zip)

**Comment by sbordet** [comment]
@andresluuk the war file you attached contains other classes that seem to be related to multipart handling.

However, if you make a request for `/image.jpg` with `Range: bytes=1100-2300,36000-41000,45012-45099` you won't hit those classes, but just hit Jetty's `DefaultServlet` to serve the static file.

Is what I describe above the scenario you expect?
Or your request is supposed to be handled your classes?

**Comment by andresluuk** [comment]
Yes if you hit image directly then those classes will not be hit. The classes there are to proxy the request into the image.
I uploaded my entire test case, I did not filter out the excess files (there should not be many). If you hit the proxy servlet (/proxy), it does multiple requests into the image with different ranges, but for some reason the simpler ranges do not get stuck.

If you deploy the application and invoke the proxy servlet, then it will make HttpURLConnection into the image (inside the PartialRequestTests file, I included the source for it also in the war, so no decomple is needed for that).
Now this request inside sets the property with Range: bytes=1100-2300,36000-41000,45012-45099 and the test then itself timeouts. But if you check the active threads on jetty after the timeout you will see the stuck thread I mentioned in the report.

So in my case I got the stuck thread if I made the request into the image from HttpURLConnection. I am not sure how todo from browser (ie without the proxy servlet), so in my request the issue happened if I made the rangerd request from java.

**Comment by sbordet** [comment]
I just wrote a test case that shows that `multipart/byteranges` requests are properly served to the client, both using Jetty's `HttpClient` and `URLConnection`, on Linux.

I am not sure I understand what you are doing with the proxying, but without source code I cannot tell -- I read `PartialRequestTests` and looks ok.

The blockage happens inside the JDK, so there is not much that Jetty can do, and seems weird that the file system blocks on a simple file read.

Bottom line, Jetty 12 seems to work fine from my testing, and this blockage happens at the file system level, and there is no apparent reason as to why that happens.

If you hit the image directly with a range request, does it block?

**Comment by andresluuk** [comment]
> I am not sure I understand what you are doing with the proxying
the proxying is just a random servlet to make a new request with PartialRequestTests, there is nothing else there.

I will try to simplify the case tomorrow morning, so that there is just 1 file remaining and send it again.
Because I am on windows it might be an windows only issue on jetty.. I made the case because it was OK with older jettys.

**Comment by sbordet** [comment]
@andresluuk definitely thanks for this report, it's appreciated and we looked into this because you reported it was working with older Jetty releases.

We just need to iterate to find what's different. We will try on Windows too.

**Comment by sbordet** [comment]
Tried on Windows too, my test (`DistributionTests.testRangeRequestMultiPartRangeResponse()`) passes cleanly.
https://github.com/eclipse/jetty.project/blob/jetty-12.0.x/tests/test-distribution/test-distribution-common/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java#L1478

**Comment by andresluuk** [comment]
[pkg-http-partialContentSimple.zip](https://github.com/eclipse/jetty.project/files/12082064/pkg-http-partialContentSimple.zip)
I made a shorter test. Removed all the exces code. Added the one java source file also inside next to the class file.
Deploy it on windows on jetty11 and make a request:
http://localhost:8080/pkg-http-partialContentSimple/proxy?685733513284632
Request passes and prints out some stuff.
Now do the same in jetty12 and it gets stuck.
I also added some debug logging that will print out all the stacks after 40 sec.

> If you hit the image directly with a range request, does it block?
Don't know exactly how todo it, but I hope you can reproduce it with my simpler example.

**Comment by andresluuk** [comment]
Another difference with your test is that your ranges are smaller.
For example in my own testing with the same image file a range of 1000-2000,3000-4000 passed also, but the bigger range mentioned in this case fails.

**Comment by sbordet** [comment]
> Another difference with your test is that your ranges are smaller.

Because the resource is smaller. Should not make any difference.

**Comment by sbordet** [comment]
Found the issue, a typical PEBKAC 🙄 
Standby for a fix.

**Comment by sbordet** [comment]
@andresluuk can you try #10125?

**Comment by sbordet** [comment]
@andresluuk we could reproduce and the fix should solve the issue.

Let us know if it works for you too.

**Comment by andresluuk** [comment]
I can confirm, that my original test passes now. Thank you.
